### PR TITLE
fix: guard against null output in parse_response and output_text property

### PR DESCRIPTION
**Pull request by adhavan18**
Fixes #3325 and #3063.

**Problem**

Two crashes in the Responses API when the server returns unexpected null values:

1. Some backends send `output: null` in the final `response.completed` event even when output items were already streamed. `parse_response` and `Response.output_text` both iterate `response.output` directly → `TypeError: 'NoneType' object is not iterable`.
2. When an output_text content item has `text: null`, `parse_response` passes `None` to the structured parser and `output_text` concatenates `None` into the string.

**Fix**

- `_parsing/_responses.py` — iterate `response.output or []`; skip structured parsing when `item.text is None`
- `types/responses/response.py` — same `or []` guard in the `output_text` property; skip content items where `text is None`
- `lib/streaming/responses/_responses.py` — when `response.completed` arrives with `output: null`, fall back to the accumulated snapshot output instead of parsing an empty list

**Test**

All existing tests in `tests/lib/responses/` pass.

### Known limitation: strict response validation

This fixes the default, non-strict path. Clients using `_strict_response_validation=True` are unaffected, because `validate_type()` rejects the payload earlier: `Response.output` is typed `List[ResponseOutputItem]` and `ResponseOutputText.text` is typed `str`, both in Stainless-generated files. Allowing `None` there needs an OpenAPI spec change rather than an edit in this repo.

The `output_text` guard does live in the generated `src/openai/types/responses/response.py`. Per CONTRIBUTING that is supported: manual modifications are persisted between generations, though they can conflict with generator changes. So no restructuring is proposed here.

### CODE REVIEW

File: src/openai/lib/_parsing/_responses.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def parse_response(
 ) -> ParsedResponse[TextFormatT]:
     output_list: List[ParsedResponseOutputItem[TextFormatT]] = []
 
-    for output in response.output:
+    for output in response.output or []:
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/openai/lib/_parsing/_responses.py:61]: This was accurate against the commit it was written on, and it is the reason the streaming file is in this PR at all. Addressed in `bbd1830`, pushed about an hour and a half after this comment.

`accumulate_event` now patches the completed event with the accumulated snapshot before parsing, rather than letting `parse_response` iterate a null output:

```python
completed_response = event.response
if completed_response.output is None and snapshot is not None:
    completed_response = build(
        type(completed_response),
        **{**completed_response.to_dict(), "output": [item.to_dict() for item in snapshot.output]},
    )
self._completed_response = parse_response(
    text_format=self._text_format,
    response=completed_response,
    input_tools=self._input_tools,
)
```

So on a null completed payload the final `ParsedResponse` carries the streamed items rather than `[]`, which is what you asked for. `build` is imported at the top of that module, and `tests/lib/responses/test_null_output.py` covers the case.

The `response.output or []` guard you commented on stays as the last line of defence for the non-streaming path, where there is no snapshot to fall back to.

File: tests/lib/responses/test_null_output.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+"""Regression tests for null-output edge cases in the Responses API."""
+from __future__ import annotations
+
+import pytest
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply tests/lib/responses/test_null_output.py]: Fixed in `de0ac7b`. You were right, and it was worse than just the unused import: `ruff check` also flagged three `I001` import-order errors in the same file.

```
tests/lib/responses/test_null_output.py:
   2:1  I001  Import block is un-sorted or un-formatted
   4:8  F401  `pytest` imported but unused
  52:5  I001  Import block is un-sorted or un-formatted
  83:5  I001  Import block is un-sorted or un-formatted
Found 4 errors.
```

`ruff check .` across the repo is clean now.

File: src/openai/types/responses/response.py [hunk 3]
```diff
@@ -479,10 +479,10 @@ def output_text(self) -> str:
         If no `output_text` content blocks exist, then an empty string is returned.
         """
         texts: List[str] = []
-        for output in self.output:
+        for output in self.output or []:
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/openai/types/responses/response.py:482]: Confirmed, and I do not think it can be fixed from this PR.

`output` is declared non-optional here:

```python
output: List[ResponseOutputItem]
```

and this file opens with:

```
# File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
```

So making the field accept `None` means changing the OpenAPI spec, which is not something a contributor can do in this repository. The same applies to the sibling comment about `ResponseOutputText.text`.

That leaves this PR fixing the non-strict path, which is the default, while `_strict_response_validation=True` still raises earlier in `validate_type()`. That is a genuine remaining gap and I would rather it be stated in the PR than discovered later, so I have noted it in the description.

Worth flagging to a maintainer: this PR does edit the generated `response.py` to guard the `output_text` property. If that file is regenerated without the change being carried in the Stainless config, the guard will be dropped. If that is a concern, the property guard could move to a non-generated module instead, and I am happy to restructure it that way.

File: src/openai/types/responses/response.py [hunk 4]
```diff
@@ -479,10 +479,10 @@ def output_text(self) -> str:
...
             if output.type == "message":
                 for content in output.content:
-                    if content.type == "output_text":
+                    if content.type == "output_text" and content.text is not None:
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/openai/types/responses/response.py]: Same root cause as the `output` comment above: `ResponseOutputText.text` is typed `str` in a Stainless-generated file, so a strict client rejects `text: null` in `validate_type()` before reaching this guard. Changing that field needs an OpenAPI spec change rather than an edit here.

The guard in this PR still covers the default non-strict path, where `text: null` currently concatenates `None` into the aggregated string.

File: src/openai/lib/streaming/responses/_responses.py [hunk 5]
```diff
@@ -357,9 +357,23 @@ def accumulate_event(self, event: RawResponseStreamEvent) -> ParsedResponseSnaps
...
+            # events and accumulated into `snapshot.output`.  In that case we
+            # must not let parse_response iterate over null and produce an empty
+            # output list; instead we patch the completed event's response with
+            # the accumulated snapshot output before parsing so that the final
+            # ParsedResponse contains the real content.
+            completed_response = event.response
+            if completed_response.output is None and snapshot is not None:
+                completed_response = build(
+                    type(completed_response),
+                    **{**completed_response.to_dict(), "output": [item.to_dict() for item in snapshot.output]},
```
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/openai/lib/streaming/responses/_responses.py]: Correct, and worth separating from this change: the accumulator has never consumed `response.output_item.done` or `response.content_part.done`. That is true on `main` today, independently of this PR:

```
$ git show main:src/openai/lib/streaming/responses/_responses.py | grep -c 'output_item\.done\|content_part\.done'
0
```

`accumulate_event` handles `output_item.added`, `content_part.added`, the two delta events, and `completed`. So a snapshot built from those can carry `in_progress` statuses and miss annotations or logprobs regardless of what this PR does.

What changes here is only what happens when `completed` arrives with `output: null`. Today that yields `output=[]` and every streamed item is lost. With the fallback it yields the accumulated items, which may have stale status. Partial data with a stale status seems clearly better than no data, so I would rather land this and treat "consume the done events in the accumulator" as its own change, since it affects every stream and not just the null-completed case.

Happy to open a separate issue for it, or to fold it in here if you would prefer one PR.

**Comment by adhavan18** [comment]
Status summary, since this has picked up a few review rounds.

**Everything raised is now answered or fixed.**

| Finding | State |
|---|---|
| P1 streamed output discarded on null completed payload | Fixed in `bbd1830`, the day it was raised |
| P2 unused `pytest` import breaks lint | Fixed in `de0ac7b` |
| P2 strict validation rejects `output: null` | Answered: needs an OpenAPI spec change |
| P2 strict validation rejects `text: null` | Answered: same root cause |
| P2 snapshot can hold stale items | Answered: pre-existing, proposed as a separate change |

The lint break was real and I had missed it. `ruff check` was reporting `F401` plus three `I001` errors in the new test file, which fails `scripts/lint` before the tests run. That is fixed and `ruff check .` is clean across the repo. `pytest tests/lib/responses/test_null_output.py` is 4 passed, and the branch is up to date with `main`.

**One decision I would like your call on before this goes further.**

The `output_text` guard in this PR lives in `src/openai/types/responses/response.py`, which is generated:

```
# File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
```

If that file is regenerated and the change is not carried in the Stainless config, the guard is silently dropped and the bug returns with no failing test to catch it. Two options, and I am happy to do either:

1. **Leave it as is**, and carry the guard in the Stainless config on your side.
2. **Move it out of the generated file**, so nothing in this PR touches generated code. `parse_response` and the streaming accumulator changes already live in non-generated modules, so this would only affect the `output_text` property.

I lean towards option 2 because it survives regeneration without anyone needing to remember, but it is your codebase and your generation setup, so I would rather ask than guess.

Related note for the same reason: the strict-validation gap the reviewer raised cannot be closed from this PR at all, since `Response.output` and `ResponseOutputText.text` are both generated as non-optional. This PR fixes the default non-strict path; strict clients would still need a spec change. That is now written up in the PR description so it does not get lost.

**Comment by adhavan18** [comment]
Correction to my previous comment: disregard the question about moving the `output_text` guard out of the generated file. I was wrong, and CONTRIBUTING says so plainly:

> Most of the SDK is generated code. Modifications to code will be persisted between generations, but may result in merge conflicts between manual patches and changes from the generator. The generator will never modify the contents of the `src/openai/lib/` and `examples/` directories.

So the guard is not at risk of being silently dropped; manual patches persist. The worst case is a merge conflict when the generator touches the same lines, which is a normal maintenance cost rather than a correctness risk. No decision is needed from you on that point and no restructuring is proposed. I have corrected the PR description too.

The rest of the previous comment stands: the lint break is fixed, `ruff check .` is clean, the tests pass, and the strict-validation limitation still needs an OpenAPI spec change rather than anything in this PR.

File: src/openai/types/responses/response.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (adhavan18) [reply src/openai/types/responses/response.py:482]: Correcting one thing I said here: I claimed the guard could be dropped on regeneration. That is wrong. CONTRIBUTING states that manual modifications are persisted between generations, so the only cost is a possible merge conflict with generator changes, not a silent revert.

The substantive point stands unchanged: `Response.output` is generated as non-optional, so a strict client still fails in `validate_type()` before reaching this guard, and closing that gap needs an OpenAPI spec change rather than an edit here.

**Review by jbeckwith-oai — CHANGES_REQUESTED** [review]
The null-output fallback still returns stale streaming items. `ResponseStreamState.accumulate_event()` only appends the payload from `response.output_item.added`; it never applies `response.output_item.done` (nor `response.content_part.done`) to the snapshot. Therefore, when the final `response.completed.response.output` is null, this branch serializes the earlier in-progress item rather than the authoritative done-event payload. A minimal `created -> output_item.added(status=in_progress) -> output_item.done(status=completed) -> completed(output=null)` stream makes `get_final_response().output[0].status` remain `in_progress`. This also makes the new comment claiming that done events are accumulated inaccurate.

Please update the accumulator/fallback to preserve the authoritative done-event item/part payloads, and add a streaming regression test that drives a null-output completed event after done events and verifies the final response (including status and final content/metadata). The direct parsing tests currently added do not exercise this new streaming path.

**Comment by adhavan18** [comment]
Thanks, you were right on both counts and I was wrong to argue this belonged in a separate change. Fixed in `dff8a92`.

**The accumulator now applies the done events.** `accumulate_event` only ever appended what `output_item.added` delivered, so the snapshot kept the in-progress item; `output_item.done` and `content_part.done` were never applied. Both are now handled, replacing the accumulated item or part with the authoritative payload.

**The comment was inaccurate and is corrected.** It claimed items were "delivered via `output_item.done` events and accumulated into `snapshot.output`", which was simply not true. It now says the done events handled above are what make the snapshot authoritative.

**Streaming regression test added**, in `tests/lib/responses/test_null_output_streaming.py`. It drives the exact sequence you described through `handle_event`, not `parse_response`, so it exercises the streaming path the earlier tests missed:

```
created -> output_item.added(status=in_progress) -> output_item.done(status=completed) -> completed(output=null)
```

and asserts the final status, text and annotations. On the previous commit the two new cases fail with exactly the symptom you predicted:

```
assert item.status == "completed"
E   AssertionError: assert 'in_progress' == 'completed'

assert len(content.annotations) == 1
E   AssertionError: assert 0 == 1
```

The annotation case is there because `content_part.done` is the only event that carries annotations; the deltas never do. A third case pins that a non-null `completed` event is still used as-is, so the fallback cannot start shadowing the normal path.

`pytest tests/lib/responses/` is 14 passed. ruff, mypy and pyright are all clean on the changed files.

One thing I have deliberately not changed, and I would rather flag it than leave it implicit: the strict-validation gap from the earlier review is still open. `Response.output` and `ResponseOutputText.text` are generated as non-optional, so a client with `_strict_response_validation=True` still fails in `validate_type()` before any of this runs. That needs an OpenAPI spec change rather than anything in this PR, and it is written up in the description.

**Comment by adhavan18** [comment]
@jbeckwith-oai this is addressed in `dff8a92`, pushed the morning after your review. flagging it because the review is still in *changes requested*, so the PR is blocked on a re-review rather than on code.

you were right, and the fix is the one you described: `accumulate_event` now applies `response.output_item.done` and `response.content_part.done` to the snapshot, so the null-output fallback serialises the authoritative done payload instead of the earlier in-progress one. the comment that claimed done events were already accumulated was wrong and is gone.

your exact repro is now a test, `tests/lib/responses/test_null_output_streaming.py`:

```
created -> output_item.added(status=in_progress) -> output_item.done(status=completed) -> completed(output=null)
  => final.output[0].status == 'completed'   (was 'in_progress')
```

three cases in there: that one, a `content_part.done` case pinning that annotations which only arrive on the done event survive, and a control that a non-null `completed` payload is still used as-is.

happy to rebase if it has gone stale.

**Comment by adhavan18** [comment]
Status check on the open threads here, since it's been a few weeks.

## @jbeckwith-oai's review: fixed

`dff8a92` addresses the stale-snapshot issue directly: `accumulate_event` now applies `response.output_item.done` and `response.content_part.done` to the snapshot (previously only `.added` was accumulated, which carries `status=in_progress` and none of the done-event-only fields like annotations). Added `tests/lib/responses/test_null_output_streaming.py`, which drives the exact repro from the review (`created -> output_item.added(in_progress) -> output_item.done(completed) -> completed(output=null)`) and fails on the parent commit with `assert 'in_progress' == 'completed'` before the fix. A third case pins that a non-null `completed` event still passes through unmodified.

## Still open: two codex findings this doesn't cover

Codex flagged (Aug 2) that a client with `_strict_response_validation=True` never reaches any of these guards, because `Response.output` is `List[ResponseOutputItem]` and `ResponseOutputText.text` is `str` — both non-Optional — so `validate_type()` in `_base_client.py:695` rejects a null payload before construction, let alone before `output_text`'s `self.output or []` check runs.

Fixing that properly means one of: widening those two field types to `Optional`, or normalizing the raw payload before `validate_type()` — and `_process_response_data` is shared by every resource in the SDK, not just Responses, so neither is a change I want to make unscoped inside this PR without a steer. Flagging it here rather than guessing, since it's a real gap for strict-validation users but a separate, bigger decision than "guard against null in the properties we already touch."

Happy to open it as its own PR either way once there's a preference — Optional fields (more correct, touches the public type) vs. normalize-on-parse (narrower blast radius, only benefits `_strict_response_validation`).

**Comment by adhavan18** [comment]
closing this, it's superseded. #3345 merged a few hours ago and fixes the same underlying bug (fixes #3325) more completely than this PR did, it also recovers finalized response.output_item.done items on null/missing output, not just guards the crash. checked the merged diff against src/openai/lib/_parsing/_responses.py and src/openai/lib/streaming/responses/_responses.py, the null-output crash this PR addresses is fixed there.

thanks for the reviews along the way @jbeckwith-oai

**Comment by adhavan18** [comment]
correction to my close a few minutes ago: i was wrong, reopening.

#3345 fixed parse_response() and the streaming accumulator, but response.output_text still crashes on the exact same bug. checked main directly:

    def output_text(self) -> str:
        texts: List[str] = []
        for output in self.output:   # still unguarded

reproduces: TypeError: 'NoneType' object is not iterable

rebuilt this pr against current main, scoped to just what's still needed since #3345 already covers parse_response and the streaming path. pushed c1006600:

- guard output.output_text the same way #3345 guarded parse_response: self.output or []
- also guard content.text being None on an output_text block (a partial/aborted item), which crashes the str.join the same way, found this while writing the test
- 3 new tests in test_null_output.py, confirmed they fail on unfixed main and pass here
- full test_null_output.py suite: 19 passed

sorry for the churn

**Comment by adhavan18** [comment]
closing, this is fully superseded now. df6b72c (`fix(responses): handle null text in output_text`, #3019, merged today) landed the identical content.text-is-none guard this pr adds, plus broader parametrized test coverage than what i had here. verified by reading the actual merged diff, not just the title this time.

the output=none guard is untouched and still in place from #3345, so nothing here is still needed.
